egs/mini_librispeech/s5/local/grammar/simple_demo.sh

```diff
@@ -0,0 +1,169 @@
+    exit 1
+  fi
+fi
+
+if [ $stage -le 4 ]; then
+  # Create the graph for the nonterminal in data/lang_nosp_grammar2b
+  # Again, we don't choose to put these symbols on the output side, but it would
+  # be possible to do so.
+  lang=data/lang_nosp_grammar2b
+  cat <<EOF | fstcompile --isymbols=$lang/words.txt --osymbols=$lang/words.txt > $lang/G.fst
+0    1    #nonterm_begin <eps>
+1    2    ONE  ONE    0.69314718055994
+1    2    TWO  TWO    0.69314718055994
+1    2    <eps>  <eps>  5.0
+2    3    #nonterm_end <eps>
+3
+EOF
+  utils/mkgraph.sh --self-loop-scale 1.0 $lang $tree_dir $tree_dir/grammar2b
+
+
+  # test that the binary 'compile-graph' does the same thing as mkgraph.sh.
+  offset=$(grep nonterm_bos $lang/phones.txt | awk '{print $2}') # 364
+  compile-graph --nonterm-phones-offset=$offset --read-disambig-syms=$lang/phones/disambig.int \
+       $tree_dir/tree $tree_dir/1.mdl $lang/L_disambig.fst $lang/G.fst $tree_dir/grammar2b/HCLG2.fst
+
+  if ! fstequivalent --delta=0.01 --random=true --npath=100 $tree_dir/grammar2b/HCLG{,2}.fst; then
+    echo "$0: two methods of producing graph in $tree_dir/grammar2b were different."
+    exit 1
+  fi
+fi
+
+if [ $stage -le 5 ]; then
+  # combine the top-level graph and the sub-graph together using the command
+  # line tools. (In practice you might want to do this from appliation code).
+
+  lang=data/lang_nosp_grammar2a
+  offset=$(grep nonterm_bos $lang/phones.txt | awk '{print $2}') # 364
+  clist=$(grep nonterm:contact_list $lang/phones.txt | awk '{print $2}') # 368
+
+  # the graph in $tree_dir/grammar2/HCLG.fst will be a normal FST (ConstFst)
+  # that was expanded from the grammar.  (we use --write-as-grammar=false to
+  # make it expand it).  This is to test equivalence to the one in
+  # $tree_dir/grammar1/
+  mkdir -p $tree_dir/grammar2
+  make-grammar-fst --write-as-grammar=false --nonterm-phones-offset=$offset $tree_dir/grammar2a/HCLG.fst \
+                   $clist $tree_dir/grammar2b/HCLG.fst  $tree_dir/grammar2/HCLG.fst
+fi
+
+if [ $stage -le 6 ]; then
+  # Test equivalence using a random path.. can be useful for debugging if
+  # fstequivalent fails.
+  echo "$0: will print costs with the two FSTs, for one random path."
+  fstrandgen $tree_dir/grammar1/HCLG.fst > path.fst
+  for x in 1 2; do
+    fstproject --project_output=false path.fst | fstcompose - $tree_dir/grammar${x}/HCLG.fst | fstcompose - <(fstproject --project_output=true path.fst) > composed.fst
+    start_state=$(fstprint composed.fst | head -n 1 | awk '{print $1}')
+    fstshortestdistance --reverse=true composed.fst | awk -v s=$start_state '{if($1 == s) { print $2; }}'
+  done
+
+fi
+
+if [ $stage -le 7 ]; then
+  echo "$0: will test equivalece using fstequivalent"
+  if fstequivalent --delta=0.01 --random=true --npath=100 $tree_dir/grammar1/HCLG.fst $tree_dir/grammar2/HCLG.fst; then
+    echo "$0: success: the two were equivalent"
+  else
+    echo "$0: failure: the two were equivalent"
+  fi
```

Comment:
Cool. In my understanding, we might need to further verify it by a complexer example (e.g. unk word recognition). My experience (from the previous AFST idea) is when we have a complexer top-level graph (e.g. the following one), I fail in making the grammar1 and grammar2 equivalent and also obtain some WER drops if decoding.
0    1    GROUP   GROUP
1    2    #nonterm:contact_list  <eps>
2    3    #nonterm:phone_number  <eps>
3    4    ASSIST   ASSIST  0.69314718055994
4    5    #nonterm:contact_list  <eps>
5

Reply:
@DongjiGao, when you make your more thorough test example, make sure to have nonterminals following each other directly, and also invoking each other.  (Not recursively though).